Frameworks/Foundation/NSScanner.mm

```diff
@@ -575,118 +557,283 @@ - (BOOL)scanUpToString:(NSString*)string intoString:(NSString**)intoString {
+        _scanLocation = i;
         return YES;
-    } else {
-        _location = oldLocation;
-
-        IwFree(result);
-        return NO;
     }
+
+    return NO;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (BOOL)scanDecimal:(NSDecimal*)decimalValue {
-    UNIMPLEMENTED();
-    return StubReturn();
+    // TODO #2369: This could probably be made more precise if it scanned directly from string -> NSDecimalNumber
+    // but WinObjC does not have [NSDecimalNumber decimalNumberWithString:] yet
+    double doubleValue;
+    if (![self scanDouble:&doubleValue]) {
+        return NO;
+    }
+
+    if (decimalValue) {
+        *decimalValue = [[NSDecimalNumber numberWithDouble:doubleValue] decimalValue];
+    }
+
+    return YES;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (BOOL)scanHexDouble:(double*)result {
-    UNIMPLEMENTED();
-    return StubReturn();
+    enum {
+        STATE_START,
+        STATE_SIGN,
+        STATE_ZERO,
+        STATE_MANTISSA,
+        STATE_P,
+        STATE_EXPONENT,
+    } state = STATE_START;
+
+    bool mantissaIsNegative = false;
+    bool exponentIsNegative = false;
+    bool scannedPeriod = false;
+    bool hasValue = false;
+
+    double mantissa = 0;
+    int exponent = 0;
+
+    // Once a '.' has been scanned, this /= 16 every character, so that each base-16 decimal digit is appropriately placed
+    // ie: 1.1ae = 1 + (1 * (mantissaDigitPlace /= 16)) + (a * (mantissaDigitPlace /= 16)) + (e * (mantissaDigitPlace /= 16))
+    double mantissaDigitPlace = 1;
+
+    const NSUInteger length = _string.length;
+    NSUInteger i = [self _indexOfNextUnskippedCharacter];
+    for (; i < length; ++i) {
+        unichar unicode = (unichar)[_string characterAtIndex:i];
+
+        switch (state) {
+            case STATE_START:
+                if (unicode == '-') {
+                    state = STATE_SIGN;
+                    mantissaIsNegative = true;
+                } else if (unicode == '+') {
+                    state = STATE_SIGN;
+                } else if (unicode == '0') {
+                    state = STATE_ZERO;
+                } else {
+                    return NO;
+                }
+                break;
+
+            case STATE_SIGN:
+                if (unicode == '0') {
+                    state = STATE_ZERO;
+                } else {
+                    return NO;
+                }
+                break;
+
+            case STATE_ZERO:
+                if (unicode == 'x' || unicode == 'X') {
+                    state = STATE_MANTISSA;
+                } else {
+                    return NO;
+                }
+                break;
+
+            case STATE_MANTISSA:
+                if (unicode == '.') {
+                    if (!scannedPeriod) {
+                        scannedPeriod = true;
+                    } else {
+                        goto done_scanning_hex_double; // Stop scanning once a second period has been scanned
```

Comment:
Since break inside a switch in a for loop breaks the switch and not the outer loop, this unfortunately seemed to be the best out of a group of bad options:
- Add a done bool that is checked on every iteration: this causes ++i to occur once more than necessary unless it's the end of the string, making the set of _scanLocation = 1 at the end more complicated than necessary
- Move the switch to a helper function: would need to manipulate both mantissa and exponent through out-pointers, seems unnecessarily messy
- Add a continue to every 'good' case, then add a break outside the switch that the 'bad' cases can fall too: wwwwaayyy too messy